boards: nrf54h20dk: enable USB device controller

jfischer-no opened the pull request:
~~Based on #71487 and #70245~~.
Rework controller initialization and vendor quirks, enable nRF54H20 USB device controller

jfischer-no committed:
drivers: udc_dwc2: rework controller initialization

Move most of the controller initialization to a separate function called
during udc_enable(). This allows us to add support for the platform
where the device controller is only available when VBUS is present and
the PHY is powered.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
dts: nordic: update USBHS node

Add "nordic,nrf-usbhs" vendor compatible and new required properties.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
boards: nrf54h20dk: enable USB device controller

Enable USB device controller.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: udc_dwc2: use devicetree to configure endpoint capabilities

Although we can get the number of configured OUT and IN endpoints and
endpoint capabilities from the DWC GHWCFGn registers, we need to
configure the number of endpoint configuration structs at build time. On
some platforms, we cannot access the hardware register at pre-init, so
we use the GHWCFGn values from the devicetree to provide endpoint
capabilities. This can be considered a workaround, and we may change the
upper layer internals to avoid it in the future.

Also, add a new vendor quirk to fill in platform-specific controller
capabilities.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: udc_dwc2: add vendor quirks to support Nordic USBHS controller

Add vendor quirks to support Nordic USBHS controller.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: udc_dwc2: rework vendor quirks

Rework and rename vendor quirks to better reflect where they intended to
be called. Number of quirks probably not final and will be trimmed
later.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -82,6 +76,8 @@ struct udc_dwc2_data {
 	/* Number of OUT endpoints including control endpoint */
 	uint8_t outeps;
 	uint8_t setup[8];
+	uint16_t numof_in_eps;
+	uint16_t numof_out_eps;
```
Why `uint16_t`? The maximum value here is 16 and therefore `uint8_t` is more than enough.

jfischer-no replied:
It was some leftover, removed it.

tmon-nordic commented on drivers/usb/udc/udc_dwc2_vendor_quirks.h:
```diff
@@ -79,6 +79,46 @@ static inline int stm32f4_fsotg_disable_phy(const struct device *dev)
+	const struct udc_dwc2_config *config = dev->config;
+	int err;
+
+	/* Just assume it is always continuous for that platform. */
```
I believe `this platform` (instead of `that platform`) would be appropriate here.

tmon-nordic commented on drivers/usb/udc/udc_dwc2_vendor_quirks.h:
```diff
@@ -148,6 +149,180 @@ DT_INST_FOREACH_STATUS_OKAY(QUIRK_STM32F4_FSOTG_DEFINE)
+	data->caps.hs = true;
+
+	/* Register nRF54H20 OUT endpoints 1 ... 5 and 12 ... 15 */
+	for (size_t i = 1; i < config->num_out_eps; i++) {
+		config->ep_cfg_out[i].caps.out = 1;
+		config->ep_cfg_out[i].caps.bulk = 1;
+		config->ep_cfg_out[i].caps.interrupt = 1;
+		config->ep_cfg_out[i].caps.iso = 1;
+		config->ep_cfg_out[i].caps.mps = 1024U;
+		if (i < 6) {
+			config->ep_cfg_out[i].addr = USB_EP_DIR_OUT | i;
+		} else {
+			config->ep_cfg_out[i].addr = USB_EP_DIR_OUT | (i + 6);
+		}
+
+		err = udc_register_ep(dev, &config->ep_cfg_out[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	/* Register nRF54H20 IN endpoints 1 ... 11 */
+	for (size_t i = 1; i < config->num_in_eps; i++) {
+		config->ep_cfg_in[i].caps.in = 1;
+		config->ep_cfg_in[i].caps.bulk = 1;
+		config->ep_cfg_in[i].caps.interrupt = 1;
+		config->ep_cfg_in[i].caps.iso = 1;
+		config->ep_cfg_in[i].caps.mps = 1024U;
+		config->ep_cfg_in[i].addr = USB_EP_DIR_IN | i;
+
+		err = udc_register_ep(dev, &config->ep_cfg_in[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	return 0;
+}
```
I believe it would be more appropriate for the quirk to simply return `GHWCFG1` value (here: `USBHSCORE_GHWCFG1_ResetValue`) and make the stack to configure the endpoint directions based on it.

jfischer-no replied:
We can add required GHWCFGn properties to the bindings and then have it generic for all the vendor implementations.

tmon-nordic replied:
That would work as well.

tmon-nordic commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -333,10 +333,12 @@
 				reg = <0x86000 0x1000>, <0x2f700000 0x40000>;
 				reg-names = "wrapper", "core";
 				interrupts = <134 NRF_DEFAULT_IRQ_PRIORITY>;
+				num-in-eps = <12>;
```
This is not exactly right. While nRF54H20 has the possibility to use 12 IN endpoints, it can only support EP0 plus any combination of up to 7 different IN endpoints (see `INEps` in ` GHWCFG4`). The stack expects that all endpoints can be used simultaneously, which is not the case here.

I would go with 8 here, possibly with a comment that it is limited due to number available of TxFIFOs.

tmon-nordic commented on boards/nordic/nrf54h20dk/Kconfig.defconfig:
```diff
@@ -7,6 +7,9 @@ choice BT_HCI_BUS_TYPE
 	default BT_HCI_IPC if BT
 endchoice
 
+config NRFS
+	default y if USB_DEVICE_STACK_NEXT
```
Why not 
```
	select NRFS if NRFS_HAS_VBUS_DETECTOR_SERVICE
	select NRFS_VBUS_DETECTOR_SERVICE_ENABLED if NRFS_HAS_VBUS_DETECTOR_SERVICE
```
under `config UDC_DWC2` in `drivers/usb/udc/Kconfig.dwc2` instead?

The selects have the advantage that `NRFS` cannot be disabled if `UDC_DWC2` is enabled. With just this default, the `NRFS` can be disabled leading to build time failure.

jfischer-no replied:
Changed as suggested.

carlescufi commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -82,6 +76,8 @@ struct udc_dwc2_data {
 	/* Number of OUT endpoints including control endpoint */
 	uint8_t outeps;
 	uint8_t setup[8];
+	uint16_t numof_in_eps;
+	uint16_t numof_out_eps;
```
Please remove

jfischer-no committed:
drivers: udc_dwc2: prevent access to registers if USBHS is not ready

On USBHS, we cannot access the DWC2 register until VBUS is detected and
valid. Kernel event API is used to block if a valid VBUS signal is not
present when the user tries to force usbd_enable().

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>
Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: udc_dwc2: fix interpretation of NUMDEVEPS and INEPS fields

The NUMDEVEPS field provides the number of endpoints in addition to the
control endpoint. It is used to iterate over GHWCFG1 register value to
get correct number of configured IN/OUT endpoints. To get it correctly,
we need to use it internally as number including control endpoint.

Interpretation of INEPS misses +1 because value 0 means 1 IN endpoint
and so on.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic reviewed (changes requested):
`CONF_FILE=usbd_next_prj.conf EXTRA_DTC_OVERLAY_FILE="ramdisk.overlay" west build --pristine -b nrf54h20dk/nrf54h20/cpuapp -d build-usb-msc-54h20-pristine ~/zephyrproject/zephyr/samples/subsys/usb/mass/ -- -DCONFIG_APP_MSC_STORAGE_RAM=y -DCONFIG_UDC_BUF_POOL_SIZE=8192`

results in
```
[00:00:23.365,309] <err> os: ***** BUS FAULT *****
[00:00:23.365,329] <err> os:   Precise data bus error
[00:00:23.365,332] <err> os:   BFAR Address: 0x14
[00:00:23.365,342] <err> os: r0/a1:  0x00000000  r1/a2:  0x88000000  r2/a3:  0x2f011604
[00:00:23.365,347] <err> os: r3/a4:  0x00000018 r12/ip:  0x2f02bfac r14/lr:  0x0e0ad5bb
[00:00:23.365,349] <err> os:  xpsr:  0x81000000
[00:00:23.365,352] <err> os: Faulting instruction address (r15/pc): 0x0e0ad5ba
[00:00:23.365,368] <err> os: >>> ZEPHYR FATAL ERROR 25: Unknown error on CPU 0
[00:00:23.365,376] <err> os: Current thread: 0x2f012188 (unknown)
[00:00:23.433,334] <err> os: Halting system
```

when running USB2CV MSC tests.

The faulting address is `tmp->stat.enabled` access in `dwc2_unset_unused_fifo()`:
https://github.com/zephyrproject-rtos/zephyr/blob/6a265e691fdb992cc934d7b629e6b1f6af904b03/drivers/usb/udc/udc_dwc2.c#L946-L954

`tmp` is NULL for `i = 8`.

tmon-nordic commented on drivers/usb/udc/udc_dwc2_vendor_quirks.h:
```diff
@@ -172,6 +181,11 @@ static inline int usbhs_enable_core(const struct device *dev)
+		LOG_WRN("VBUS is not ready, block udc_enable()");
+	}
+
+	k_sem_take(&vbus_ready_sem, K_FOREVER);
```
This is good enough for first enable. However if application does `udc_enable()`, `udc_disable()`, `udc_enable()` while the USB cable is connected just once and held in place, then the code will block claiming VBUS is not ready.

I think `k_event` might be more appropriate here. `usbhs_vbus_handler()` would be posting and clearing the event bit while `usbhs_enable_core()` would simply call `k_event_wait()` without event reset.

tmon-nordic replied:
Yes, we need to track the VBUS state and hence the event API is best suited for this action. Please check the implementation at https://github.com/jfischer-no/zephyr/pull/4

jfischer-no commented on drivers/usb/udc/udc_dwc2_vendor_quirks.h:
```diff
@@ -107,6 +108,142 @@ DT_INST_FOREACH_STATUS_OKAY(QUIRK_STM32F4_FSOTG_DEFINE)
+{
+	NRF_USBHS_Type *wrapper = USBHS_DT_WRAPPER_REG_ADDR(0);
+
+	wrapper->ENABLE = 0UL;
+	wrapper->TASKS_START = 1UL;
+
+	/* Enable interrupt */
+	wrapper->INTENSET = 0UL;
```
```
	/* Disable interrupts */
	wrapper->INTENSET = 0UL;

	wrapper->ENABLE = 0UL;
	wrapper->TASKS_START = 1UL;
```

tmon-nordic replied:
`wrapper->INTENCLR = 1UL;` is clearing. Alternatively you can use `wrapper->INTEN = 0UL;` to disable interrupt. Writing `0` to `INTENSET` has no effect.

tmon-nordic commented on drivers/usb/udc/udc_dwc2_vendor_quirks.h:
```diff
@@ -96,6 +108,158 @@ DT_INST_FOREACH_STATUS_OKAY(QUIRK_STM32F4_FSOTG_DEFINE)
+	NRF_USBHS_Type *wrapper = USBHS_DT_WRAPPER_REG_ADDR(0);
+
+	/* Disable interrupts */
+	wrapper->INTENSET = 0UL;
```
This is not disabling interrupts. It is essentially no action.

Either change it to `wrapper->INTEN = 0UL;` or `wrapper->INTENCLR = 1UL;`.

jfischer-no committed:
samples: usb: add nrf54h20dk to allowed platforms

nRF54H20DK board is first major user of new UDC DWC2 driver.
Add nRF54H20DK (Application CPU) to allowed platforms in USB device
samples.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>
